### ISSUE TITLE
Add a clonable ArcError/ArcResult

### ISSUE THREAD

**Issue by ssbr** [dtolnay/anyhow#245]
I really would like to be able to clone errors. For example, in order to memoize a function that can fail (my use case -- I want to return a `Result<T>` from a [Salsa](https://github.com/salsa-rs/salsa) query). In order to e.g. allow downcasting `anyhow::Error` by value -- which is a major reason `Error` doesn't already use reference counting -- a clonable reference counted error would need to be a new type, e.g. `ArcError`, which is to `Error` as `Arc<dyn Error>` is to `Box<dyn Error>`. 

This does need to be layered appropriately in user code and in APIs: it's difficult/expensive to move from an `Arc<dyn ...>` to a `Box<dyn ...>`, but trivial to move in the other direction. The same would presumably be true of `ArcError` and `Error`.

---

In #7 (now closed), @dtolnay [explained why `Error` doesn't implement `Clone`](https://github.com/dtolnay/anyhow/issues/7#issuecomment-539383249):

> In order to provide a Clone impl we would need to do either of:
> 
> - require that you can only construct anyhow::Error from errors that are Clone, which would make it incompatible with many error types in the wild like io::Error and serde_json::Error; or
> - reference-count the internal error.
> 
> The first one seems bad to me and the second one makes it impossible to downcast by value which also seems bad (fwiw failure came to the same conclusion in https://github.com/rust-lang-deprecated/failure/issues/148).
> 
> Is there a way for you to work around this by using Rc<Error> or Arc<Error> in the tests?

This is well reasoned, and I think solution 1 is right out. Solution 2 is fixable: if it were a new type, then we can still downcast `Error` by value, but not `ArcError`. One question left, then.

## Why not `Arc<anyhow::Error>`?

The obvious approach is to use `Arc<anyhow::Error>` and `Result<T, Arc<anyhow::Error>>` as the top level types, in place of my suggested `ArcError` type etc. This requires no changes to `anyhow` and can be done entirely within user code.

In fact, that's basically what I'm suggesting, I just think it should be in `anyhow` and not in user code. It is a lot of trouble to get working -- I tried this in https://github.com/google/crubit/blob/main/common/arc_anyhow.rs, results summarized below.

(These issues also apply to the the slightly worse alternative `Arc<dyn std::error::Error + Send + Sync + 'static>`. The main difference between the two is that `Arc<dyn ... >` requires boxing for more operations, because it needs to recreate the `Error` rather than saving it.)

### Why not literally `Result<T, Arc<Error>>` ?

`Arc<anyhow::Error>` is not actually a drop-in replacement. It requires more boilerplate -- you can no longer use this anywhere one would use `Result<T, anyhow::Error>`, because while it can be created from a `Result<T, anyhow::Error>`, it cannot be created from a `Result<T, syn::Error>` or what have you.

```rs
// before
fn foo() -> Result<String> {
  some_function()?;
  ...
}
```

```rs
// after
fn foo() -> Result<String, Arc<Error>> { Ok(foo_impl()?) }
fn foo_impl() -> Result<String, Error> {
  some_function()?;
  ...
}
```

It's also only one directional. That might be expected (`Box`->`Arc` is one-directional? Of course it is!), but it means you actually can't just stuff everything inside foo_impl: what if foo also needs to call a different function that returns an `Arc<Error>`? 

```rs
fn bar() -> Result<(), Arc<Error>> { ... }
fn foo() -> Result<String, Arc<Error>> { Ok(foo_impl()?) }
fn foo_impl() -> result<String, Error> {
  some_function()?;
  bar()?;  // doesn't compile
  ...
}
```

It's viral and troublesome. Ideally, we'd make `bar()?` compile, or even get rid of `foo_impl` entirely and let `some_function()?` compile. Both of these require a newtype wrapping `Arc<anyhow::Error>`.

### Why not `struct Error(Arc<anyhow::Error>);` and `Result<T, Error>`?

We can get conversion from `Error` to `ArcError` working, and even can get conversion from `T : std::error::Error + Send + Sync + 'static` working. This is a simple `From` impl and fixes the issues with raw `Arc`. However, we cannot get `context`:

```rs
use anyhow::Context as _;
fn foo() -> Result<(), anyhow::Error> { ... }
fn bar() -> Result<(), Arc<anyhow::Error>> { ... }
fn baz() -> Result<(), Arc<anyhow::Error>> {
  // no problems:
  foo()?;
  foo().context("...")?;
  bar()?;
  
  bar().context("...");  // does not compile
  Ok(())
}
```

This is difficult to fix: `Context` isn't implemented because it expects either a `dyn Error + ...` or an `Error` by value, whereas we have an `Arc<Error>`, which is neither. Even if we ignore this and write our own context implementation, the only way exposed to attach context to an error is the [Error::context()](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.context) method which _also_ takes the Error by value -- and here, we have an `Arc<Error>`, not an `Error`, and are once again stuck.

The obvious workaround is to convert our `Arc<Error>` into an `Error` by value. (How do you convert an `Arc` to a `Box`? Well, by having a `Box<Arc<...>>` :))

The only way I could figure out to get this working was to implement `std::error::Error` on a *second* newtype wrapper around `Arc<Error>`. It needs to be a different newtype than the one wrapping `Arc<Error>`, for the same reason that `Error` itself doesn't implement `std::error::Error`. (That is: because it would make it impossible to write a blanket impl to convert to it from any `Error` type.)  So we need one newtype around `Arc<Error>` that we use for results, and a second newtype which exists solely to convert into `anyhow::Error`.

And so to implement `Arc<Error>` with `context`, you need the following top-level items at minimum:

```rs
pub struct Error(Arc<anyhow::Error>);
impl Display for Error {...}
impl Debug for Error {...}
impl<T> From<T> for Error where T: Into<anyhow::Error> {...}
struct StdError(Arc<anyhow::Error>);
impl Display for StdError {...}
impl Debug for StdError {...}
impl std::error::Error for StdError {...}
impl From<Error> for StdError {...}
pub trait Context<T, E> {
    /* duplication of anyhow::Context, because anyhow::Context is sealed. */
}
impl<T, E> Context<T, E> for Result<T, E> where E: std::error::Error + Send + Sync + 'static {...}
impl<T> Context<T, std::convert::Infallible> for Option<T> {...}
impl<T> Context<T, Error> for Result<T, Error>  {...}
```

Optionally (to allow using context with `anyhow::Error`), also:

```rs
impl<T> Context<T, anyhow::Error> for Result<T, anyhow::Error>  {...}
```

Finally, we have to reimplement all the macros to return `ArcError`/`ArcResult`:

```rs
macro_rules! anyhow { ... }
macro_rules! ensure { ... }
macro_rules! bail { ... }
```

All told, this is a couple hundred lines of very tricky code.

There may be more things still to go, I haven't tried to progress past this point. (I think downcasting might mostly just work with a Deref impl, fortunately.)

Once again, here's a link to my actual attempt at this: https://github.com/google/crubit/blob/main/common/arc_anyhow.rs

In summary, I think it's much nicer if we just have all this work done in one place upstream, as it is not simple to create an `Arc<anyhow::Error>`.

**Comment by dtolnay** [comment]
I think this crate is all set with a single error type, so I am not interested in adding a separate clonable reference counted error type into this crate. I would recommend pursuing this in a different crate.

**Comment by ssbr** [comment]
Okay.

**Comment by ennis** [comment]
I ran into the same limitation with a similar use case (caching results of fallible functions).
